phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -333,6 +333,8 @@ private int executeTool(Connection conn, ConnectionQueryServices queryServices,
                 LOGGER.info("Executing " + operation + " for " + dataTableFullName);
 
                 boolean mutable = !(dataTable.isImmutableRows());
+                disableTable(admin, dataTableFullName, indexes);
+                modifyTable(admin, dataTableFullName, indexes);
                 if (!mutable) {
```

Comment:
Also, I think we should have a single sleep for the given batch of tables.

Reply:
Let's say there are 2 immutable tables in the list and after waiting for cache expiration for the first time and while remaining steps are in progress, the second table's GIC boolean is loaded in the cache due to some other read/write request. Now, we have not upgraded the second table yet. However, we have waited for cache to expire before and hence if we won't, now, the old GIC boolean value will be there in the cache (for second table) even though the server has changed the coproc for the second table. Does this make sense? 

On the other side of the code, is it possible to invalidate the cache after server-side values have been updated or update the cache itself via server? @gjacoby126 @priyankporwal

Reply:
That would require having all tables in the batch be disabled at once, right? So it's a tradeoff -- the process would be quicker, and the overall pain less, but the pain that's there would be greater.

Reply:
I would prefer waiting each time there is an immutable table if invalidating cache is not an option. what say?